vumi/middleware/session_length.py

```diff
@@ -42,37 +42,43 @@ def setup_middleware(self):
     def teardown_middleware(self):
         yield self.redis.close_manager()
 
+    def _generate_redis_key(self, message, address):
+        return '%s:%s:%s' % (
+            message.get('transport_name'), address, 'session_created')
+
+    @inlineCallbacks
+    def _set_redis_time(self, redis_key, time):
+        if (yield self.redis.get(redis_key)) is None:
+            yield self.redis.setex(redis_key,  self.timeout, str(time))
```

Comment:
I'm not quite sure if our current redis library supports this, but the redis command `SET key value EX timeout NX` will do this in one, atomic command. http://redis.io/commands/set